clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -287,6 +287,7 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
             this.sender = new Sender(client,
                     this.metadata,
                     this.accumulator,
+                    config.getInt(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION) == 1,
```

Comment:
It seems a little weird to treat this as a special case, though it's probably justified by common usage. However, I'm wondering if it makes sense to handle it more generically. In other words, if max-in-flight is n, then you only mute the partition if the number of in-flight batches exceeds n for that partition. That changes the meaning of the max-in-flight configuration value, but probably to something which is closer to what users have in mind when they configure it anyway. The old comment below makes me thing you may have tried this initially, but maybe you found a problem?